tech_specs/subprojects.md

```diff
@@ -0,0 +1,60 @@
+# Project division
+
+**Note:** See `service_components.txt` for diagrams.
+
+Current state of affairs is that all relevant code is spread over a set of repositories, contained in various accounts.
+
+It needs a more organized setup, but there are some considerations.
+* Functionality is logically spread over multiple loosely coupled parts:
+    * The controller firmware (C++)
+    * The controller <-> service communication layer - multiple variants (Python)
+    * (Future) plugins for external hardware not routed through the controller (Python)
+    * Various service-side functionality such as object / measurement databases (Pythyon)
+    * The Server-Side Events API (Python)
+    * The REST API (Python)
+    * The GUI (HTML/JS using the Vue/Vuex framework)
+    * (Future) Application wrappers around the GUI, making it a mobile or desktop app.
+* The long-term vision for the product includes it evolving into a generic industrial controller - without a brewery in sight.
+* Current naming themes include easily confused terms as "blocks", "blox" and "box".
+* It is very likely that future evolutions and spinoffs of the product will want to mix and match the various software components, while adding their own thing.
+
+The C++ controller firmware can continue as is: it has a well-defined interface and set of responsibilities. 
+Likewise for the GUI: it is the most specific component, and can easily be swapped out.
+
+This leaves the service layer, which currently has five separate responsibilities, some of which communicate.
+* The backend plugins (controlbox, legacy, and external) register endpoint handlers with the REST API.
+* The backend plugins publish events to the application event handler.
+* The logging database (InfluxDB) registers endpoints with the REST API.
+* InfluxDB subscribes to events published to the event handler.
+* SSE registers endpoints with the REST API.
+* SSE subscribes to events published to the event handler.
+
+REST, SSE, and the event handler are core components of the service layer: everything else can assume these are present.
+
+InfluxDB has no implementation-specific dependencies, and will simply log whatever is given to him. Data input is as simple as logging all published events. REST requests for data queries can also be directly forwarded, without needing an abstraction layer.
```

Comment:
Maybe, but influx does not know about the mapping of its keys to activities or brews (a batch of beer going through mashing, fermentation, conditioning). A rest API could provide this and fetch the data from influx. We discussed this as a reporting module, where raw influx data is combined and grouped into reports that make sense to the user.

Reply:
Described under `Splitting up` chapter, as the concept was introduced here.